tools/audio-codec-cli/src/encode.rs

```diff
@@ -14,9 +22,296 @@ use webrtc::{
+        },
+        mvex: Some(mvex),
+        traks: vec![opus_track],
+        ..Default::default()
+    };
+
+    moov.write_box(&mut writer)?;
+    writer.flush()?;
+
+    // write fragments
+    let mut fragment_sequence_number = 1;
+    // hoping the time units are in timescale...
+    let mut fragment_start_time = 0;
+
+    // use a closure to write fragments
+    let mut write_opus_frame =
+        |sample_buf: &[u8], total_length: usize, sample_lengths: Vec<u32>| -> anyhow::Result<()> {
+            let num_samples_in_trun = sample_lengths.len() as u32;
+            // create a traf and push to moof.trafs for each track fragment
+            let traf = TrafBox {
+                //  track fragment header
+                // size is 9 + header_size
+                tfhd: TfhdBox {
+                    version: BOX_VERSION,
+                    // 0x020000: default-base-is-moof is 1 and base-data-offset-present is 0
+                    // memory addresses are relative to the start of this box
+                    //
+                    // 0x10: sample size is present
+                    flags: 0x020000, //| 0x10,
+                    track_id: TRACK_ID,
+                    //default_sample_size: Some(1),
+                    ..Default::default()
+                },
+                // track fragment decode time
+                // size is 9 + header_size
+                tfdt: Some(TfdtBox {
+                    version: BOX_VERSION,
+                    flags: 0,
+                    base_media_decode_time: fragment_start_time,
+                }),
+                // track fragment run
+                // size is 13 + sample_length + header_size
+                trun: Some(TrunBox {
+                    version: BOX_VERSION,
+                    // data-offset-present, sample-size-present
+                    flags: 1 | 0x200,
+                    sample_count: num_samples_in_trun,
+                    // warning: this needs to be changed after the moof box is declared
+                    data_offset: Some(0),
+                    sample_sizes: sample_lengths,
+                    ..Default::default()
+                }),
+            };
+
+            let mut moof = MoofBox {
+                mfhd: MfhdBox {
+                    version: BOX_VERSION,
+                    flags: 0,
+                    sequence_number: fragment_sequence_number,
+                },
+                trafs: vec![traf],
+            };
+
+            let moof_size = moof.box_size();
+            if let Some(trun) = moof.trafs[0].trun.as_mut() {
+                trun.data_offset = Some(moof_size as i32 + 8);
+                //println!("trun data offset is: {:?}", trun.data_offset);
+            }
+
+            fragment_start_time += num_samples_in_trun as u64;
+            fragment_sequence_number += 1;
+
+            // this commented out code is helpful for debugging invalid mp4 files.
+            //let mfhd_size = moof.mfhd.box_size();
```

Comment:
I wonder if this could be put behind a flag so one could use it at runtime for debugging